frontend/apps/reader/modules/readerbookmark.lua

```diff
@@ -292,56 +366,262 @@ function ReaderBookmark:onShowBookmark()
+                end,
+            },
+        }
+        local second_button_row = {
+            {
+                text = from_start_text,
+                enabled = #bookmark.bookmarks > 1,
+                callback = function()
+                    UIManager:close(self.textviewer)
+                    bm_menu:onMenuHold(bookmarks[1])
+                end,
+            },
+            {
+                text = backward_text,
+                enabled = #bookmark.bookmarks > 1,
+                callback = function()
+                    UIManager:close(self.textviewer)
+                    local prev = current_bookmark - 1
+                    if prev < 1 then
+                        prev = #bookmarks
+                    end
+                    bm_menu:onMenuHold(bookmarks[prev])
+                end,
+            },
+            {
+                text = _("Close"),
+                is_enter_default = true,
+                callback = function()
+                    UIManager:close(self.textviewer)
+                    -- Also close list of bookmarks, for quick return to document. To only close the textviewer, use the close button in the upper right corner:
+                    UIManager:close(bookmark.bookmark_menu)
+                end,
+            },
+            {
+                text = forward_text,
+                enabled = #bookmark.bookmarks > 1,
+                callback = function()
+                    UIManager:close(self.textviewer)
+                    local next = current_bookmark + 1
+                    if next > #bookmarks then
+                        next = 1
+                    end
+                    bm_menu:onMenuHold(bookmarks[next])
+                end,
+            },
+            {
+                text = from_end_text,
+                enabled = #bookmark.bookmarks > 1,
+                callback = function()
+                    UIManager:close(self.textviewer)
+                    bm_menu:onMenuHold(bookmarks[#bookmarks])
+                end,
+            },
+        }
+        local text = item.notes
+        if not item.highlighted then
+            text = T(_("Page %1 %2 @ %3"), ipage, item.notes, item.datetime)
+        end
+
+        local function split(str, pat)
+            local t = {}  -- NOTE: use {n = 0} in Lua-5.0
+            local fpat = "(.-)" .. pat
+            local last_end = 1
+            local s, e, cap = str:find(fpat, 1)
+            while s do
+                if s ~= 1 or cap ~= "" then
+                    table.insert(t, cap)
+                end
+                last_end = e + 1
+                s, e, cap = str:find(fpat, last_end)
+            end
+            if last_end <= #str then
+                cap = str:sub(last_end)
+                table.insert(t, cap)
+            end
+            return t
+        end
+
+        -- HELPER FUNCTIONS WHICH PROBABLY SHOULD BE PUT SOMEWHERE ELSE:
+
+        local function substrCount(subject, needle)
```

Comment:
At least, put them at the module level, outside of (above) `function ReaderBookmark:onShowBookmark()`, so they are not re-defined each time this function is called.